IFS-102. Login on backend

uvNikita opened the pull request.

Gachapen commented on deploy/template/nginx_site.conf:
```diff
@@ -16,6 +16,17 @@ server {
         }
     }
 
+    location /auth {
```
Hm. All URLs in the API doc should be relative to the base path '/api/0'. Any reason that this isn't?

uvNikita replied:
I'm not sure about this. Because this is not an API really -- we don't get or post any data.
But I can change it to `/api/0/auth`, maybe this will be more appropriate.

Gachapen commented on server/auth/auth.py:
```diff
@@ -0,0 +1,38 @@
+        if current_user.is_active:
+            return json.dumps({'name': current_user.full_name})
+        else:
+            abort(401)
```
Maybe add a message to this and the next abort explaining what was wrong?

uvNikita replied:
Yes, you're right :)

Gachapen commented on server/auth/util.py:
```diff
@@ -0,0 +1,7 @@
+
+
+def hash_password(password):
+    return sha1(password.encode() + current_app.config['SECRET_KEY'].encode()).hexdigest()
```
Is this right? The API doc specifies SHA256.

uvNikita replied:
Yes, just took the first one to test and forgot to change

Gachapen commented on server/models.py:
```diff
@@ -6,11 +6,19 @@
     id = db.Column(db.Integer, primary_key=True)
+
+    user_name = db.Column(db.String(200), unique=True, nullable=False)
+    password = db.Column(db.String(200), nullable=False)
```
It's fine to just add this to the Lecturer Model for now I guess, but should ideally be some common model for both a student and a lecturer (and any eventual other roles), right? E.g. by extending a User class.

uvNikita replied:
Yes, more right way to do this, I think is to create `User` model with roles as `lecturer` and `student`. But for now we don't have students in our system and it was easier to do like this. We can change it in future, I think.

Gachapen commented on server/models.py:
```diff
@@ -6,11 +6,19 @@
+    user_name = db.Column(db.String(200), unique=True, nullable=False)
+    password = db.Column(db.String(200), nullable=False)
+    active = db.Column(db.Boolean, default=True)
+    admin = db.Column(db.Boolean, default=False)
```
I guess admin is OK to have for later, but there's no reason to have it no, right?

uvNikita replied:
Yes, I was thinking about admin panel. For now we don't use it, I can remove it and add if we will manage to create this panel.

Gachapen commented on server/models.py:
```diff
@@ -23,6 +31,25 @@ def full_name(self):
+        return False
+
+    def get_id(self):
+        return str(self.id)
```
What are `is_authenticated`, `is_admin`, `is_anonymous` and `get_id` for? I don't see that they are used, are some other modules using them?

uvNikita replied:
They are the part of [Flask-Login] (https://flask-login.readthedocs.org/en/latest/#your-user-class) specification. We will have anonimous user if it wasn't logged in. In future, I think,  we will be able replace our `client-id` with it.

Gachapen commented on server/tests/test_login.py:
```diff
@@ -0,0 +1,47 @@
+        db.session.flush()
+
+    def test_success(self):
+        res = self.client.post('/auth/login', data={'user_name': 'simon', 'password': '1234'})
```
Wasn't the password supposed to be hashed by the front end to prevent people sniffing the actual password?

uvNikita replied:
Yes, it will be hashed on client as well. Just think that '1234' is a hash we got from the clientside.

uvNikita commented:
I think, I fixed everything and now login should satisfy current API.
Also removed server-side password hashing as, I think, it is redundant, we will not imrove anything doing this.

Gachapen commented:
There have been some more changes to the API doc. Please look through them and update the implementation. I'll merge it when you're done.

uvNikita commented:
Done
